File: dev/web_viewer/src/models/motion/faceformer/compare_outputs.py
#!/usr/bin/env python3
"""
Compare outputs between Python FaceFormer and JavaScript FaceFormer
This script runs the Python model and saves outputs for comparison with JS model
"""

import torch
import torch.nn as nn
import numpy as np
import json
import sys
import os

# Add the FaceFormer repo to the path
sys.path.append('/home/barberb/motion/engine/FaceFormer_repo')

try:
    from faceformer import Faceformer
    from wav2vec import Wav2Vec2Model
except ImportError as e:
    print(f"❌ Could not import FaceFormer modules: {e}")
    print("Make sure the FaceFormer repository is available")
    sys.exit(1)

class Args:
    """Mock args class for FaceFormer initialization"""
    def __init__(self):
        self.dataset = "vocaset"  # or "BIWI"
        self.feature_dim = 64
        self.vertice_dim = 15069  # Number of vertices * 3
        self.period = 25
        self.device = "cpu"  # Use CPU for compatibility
        self.train_subjects = "F1 F2 F3"  # Mock subjects

def load_test_data():
    """Load the test data from JSON file"""
    test_data_path = "/home/barberb/motion/engine/web_porting_poc/faceformer/faceformer_minimal_test_data.json"
    
    try:
        with open(test_data_path, 'r') as f:
            data = json.load(f)
        return data
    except Exception as e:
        print(f"❌ Could not load test data: {e}")
        return None

def run_python_model_comparison():
    """Run the Python FaceFormer model with test data and save outputs"""
    print("🐍 Running Python FaceFormer Model Comparison...")
    
    # Load test data
    test_data = load_test_data()
    if test_data is None:
        return None
    
    # Extract inputs
    inputs = test_data["inputs"]
    audio_features = torch.tensor(inputs["audio_features"], dtype=torch.float32)
    vertice_emb = torch.tensor(inputs["vertice_emb"], dtype=torch.float32) 
    one_hot = torch.tensor(inputs["one_hot"], dtype=torch.float32)
    template = torch.tensor(inputs["template"], dtype=torch.float32)
    
    print(f"📊 Input shapes:")
    print(f"  Audio features: {audio_features.shape}")
    print(f"  Vertice embedding: {vertice_emb.shape}")
    print(f"  One-hot: {one_hot.shape}")
    print(f"  Template: {template.shape}")
    
    try:
        # Initialize the model
        args = Args()
        model = Faceformer(args)
        model.eval()
        
        print("✅ Python FaceFormer model initialized successfully")
        
        # Run a single inference step (mimicking the minimal model)
        with torch.no_grad():
            # Use the template as input (like our minimal model)
            template_reshaped = template  # Already in [1, 1, 15069] format
            
            # Create mock audio (since we can't run full wav2vec easily)
            # Use the provided audio features directly
            mock_hidden_states = audio_features  # [1, 1, 768]
            mock_hidden_states = model.audio_feature_map(mock_hidden_states)  # Map to feature_dim
            
            # Get object embedding
            obj_embedding = model.obj_vector(one_hot)  # [1, feature_dim]
            
            # Create vertice input (template difference)
            vertice_input = template_reshaped - template_reshaped  # Zero difference for first step
            vertice_input = model.vertice_map(vertice_input)  # Map to feature_dim
            
            # Add style embedding
            style_emb = obj_embedding.unsqueeze(1)  # [1, 1, feature_dim] 
            vertice_input = vertice_input + style_emb
            
            # Apply positional encoding
            vertice_input_pe = model.PPE(vertice_input)
            
            # Create masks
            tgt_mask = model.biased_mask[:, :1, :1].clone().detach()  # [heads, 1, 1]
            memory_mask = torch.zeros(1, mock_hidden_states.shape[1], dtype=torch.bool)  # [1, seq_len]
            
            # Run transformer decoder
            decoder_output = model.transformer_decoder(
                vertice_input_pe, 
                mock_hidden_states, 
                tgt_mask=tgt_mask, 
                memory_mask=memory_mask
            )
            
            # Map back to vertex space
            vertex_delta = model.vertice_map_r(decoder_output)
            
            # Add template to get final vertices
            final_vertices = vertex_delta + template_reshaped
            
            # Prepare results
            results = {
                "python_model_outputs": {
                    "new_vertice_out": final_vertices.squeeze().tolist(),
                    "updated_vertice_emb": decoder_output.squeeze().tolist(),  # Use decoder output as embedding
                    "vertex_delta": vertex_delta.squeeze().tolist(),
                    "decoder_output": decoder_output.squeeze().tolist(),
                    "style_embedding": style_emb.squeeze().tolist(),
                    "obj_embedding": obj_embedding.squeeze().tolist()
                },
                "shapes": {
                    "final_vertices": list(final_vertices.shape),
                    "decoder_output": list(decoder_output.shape),
                    "vertex_delta": list(vertex_delta.shape),
                    "style_embedding": list(style_emb.shape),
                    "obj_embedding": list(obj_embedding.shape)
                },
                "model_info": {
                    "feature_dim": args.feature_dim,
                    "vertice_dim": args.vertice_dim,
                    "period": args.period,
                    "dataset": args.dataset
                }
            }
            
            print("✅ Python model inference completed successfully")
            print(f"📐 Output shapes:")
            for key, shape in results["shapes"].items():
                print(f"  {key}: {shape}")
            
            return results
            
    except Exception as e:
        print(f"❌ Error running Python model: {e}")
        import traceback
        traceback.print_exc()
        return None

def simplified_python_model():
    """Run a simplified version that matches our minimal ONNX model"""
    print("🔬 Running Simplified Python Model (matching ONNX minimal)...")
    
    # Load test data
    test_data = load_test_data()
    if test_data is None:
        return None
    
    # Extract inputs
    inputs = test_data["inputs"]
    audio_features = torch.tensor(inputs["audio_features"], dtype=torch.float32)
    vertice_emb = torch.tensor(inputs["vertice_emb"], dtype=torch.float32) 
    one_hot = torch.tensor(inputs["one_hot"], dtype=torch.float32)
    template = torch.tensor(inputs["template"], dtype=torch.float32)
    
    # Define simplified model components matching the minimal ONNX export
    class MinimalFaceFormer(nn.Module):
        def __init__(self, audio_dim=768, embedding_dim=64, vertex_dim=15069, num_subjects=3):
            super().__init__()
            
            # Audio processing
            self.audio_proj = nn.Linear(audio_dim, embedding_dim)
            
            # Embedding processing  
            self.emb_proj = nn.Linear(embedding_dim, embedding_dim)
            
            # Subject conditioning
            self.subject_embedding = nn.Linear(num_subjects, embedding_dim)
            
            # Fusion layer
            self.fusion = nn.Linear(embedding_dim * 3, embedding_dim)
            
            # Output projections
            self.vertex_proj = nn.Linear(embedding_dim, vertex_dim)
            self.emb_update = nn.Linear(embedding_dim, embedding_dim)
            
            # Initialize weights
            self._init_weights()
        
        def _init_weights(self):
            for module in self.modules():
                if isinstance(module, nn.Linear):
                    nn.init.xavier_uniform_(module.weight)
                    if module.bias is not None:
                        nn.init.zeros_(module.bias)
        
        def forward(self, audio_features, vertice_emb, one_hot, template):
            batch_size = audio_features.shape[0]
            
            # Process inputs
            audio_proj = self.audio_proj(audio_features)  # [B, 1, emb_dim]
            emb_proj = self.emb_proj(vertice_emb)         # [B, 1, emb_dim]
            subject_emb = self.subject_embedding(one_hot) # [B, emb_dim]
            subject_emb = subject_emb.unsqueeze(1)        # [B, 1, emb_dim]
            
            # Fuse all features
            fused = torch.cat([audio_proj, emb_proj, subject_emb], dim=-1)  # [B, 1, emb_dim*3]
            fused = self.fusion(fused)  # [B, 1, emb_dim]
            
            # Generate outputs
            vertex_delta = self.vertex_proj(fused)    # [B, 1, vertex_dim]
            new_vertices = template + vertex_delta    # [B, 1, vertex_dim]
            updated_emb = self.emb_update(fused)      # [B, 1, emb_dim]
            
            return new_vertices, updated_emb
    
    # Create and run the model
    model = MinimalFaceFormer()
    model.eval()
    
    with torch.no_grad():
        new_vertices, updated_emb = model(audio_features, vertice_emb, one_hot, template)
        
        results = {
            "simplified_python_outputs": {
                "new_vertice_out": new_vertices.squeeze().tolist(),
                "updated_vertice_emb": updated_emb.squeeze().tolist()
            },
            "shapes": {
                "new_vertices": list(new_vertices.shape),
                "updated_emb": list(updated_emb.shape)
            }
        }
        
        print("✅ Simplified Python model completed")
        print(f"📐 Output shapes:")
        for key, shape in results["shapes"].items():
            print(f"  {key}: {shape}")
        
        return results

def save_comparison_data():
    """Save both model outputs for comparison"""
    print("💾 Saving comparison data...")
    
    # Run both models
    python_results = run_python_model_comparison()
    simplified_results = simplified_python_model()
    
    # Combine results
    comparison_data = {
        "timestamp": "2025-07-13",
        "test_inputs": load_test_data()["inputs"] if load_test_data() else None,
        "python_results": python_results,
        "simplified_results": simplified_results,
        "comparison_notes": {
            "python_model": "Full FaceFormer model (may have issues due to wav2vec complexity)",
            "simplified_model": "Minimal implementation matching ONNX export structure",
            "js_model": "To be compared - should match simplified_model outputs"
        }
    }
    
    # Save to file
    output_path = "/home/barberb/motion/dev/web_viewer/faceformer/python_model_outputs.json"
    try:
        with open(output_path, 'w') as f:
            json.dump(comparison_data, f, indent=2)
        print(f"✅ Comparison data saved to: {output_path}")
        return output_path
    except Exception as e:
        print(f"❌ Failed to save comparison data: {e}")
        return None

if __name__ == "__main__":
    print("🔍 FaceFormer Python vs JavaScript Output Comparison")
    print("=" * 60)
    
    output_file = save_comparison_data()
    
    if output_file:
        print("\n📋 Next steps:")
        print("1. Run the JavaScript model with the same test inputs")
        print("2. Compare the outputs using the comparison script")
        print("3. Analyze differences and adjust the models accordingly")
        print(f"\n📄 Python model outputs saved to: {output_file}")
    else:
        print("\n❌ Failed to generate comparison data")

File: dev/web_viewer/src/models/motion/faceformer/compare_outputs.js
// JavaScript Model Output Comparison Script
// Runs the JS FaceFormer model and compares with Python outputs

class FaceFormerComparison {
    constructor() {
        this.pythonResults = null;
        this.jsResults = null;
    }

    async loadPythonResults() {
        console.log('📥 Loading Python model results...');
        try {
            const response = await fetch('./python_model_outputs.json');
            this.pythonResults = await response.json();
            console.log('✅ Python results loaded successfully');
            return true;
        } catch (error) {
            console.error('❌ Failed to load Python results:', error);
            return false;
        }
    }

    async runJavaScriptModel() {
        console.log('🟨 Running JavaScript FaceFormer model...');
        
        try {
            // Initialize the JS model
            const generator = new FaceFormerWebGeneratorFixed();
            await generator.initialize();
            
            // Load test data (same as Python)
            const testDataResponse = await fetch('./faceformer_minimal_test_data.json');
            const testData = await testDataResponse.json();
            
            const inputs = testData.inputs;
            
            // Extract inputs
            const audioFeatures = inputs.audio_features[0][0];  // Flatten to 1D array
            const template = inputs.template[0][0];             // Flatten to 1D array  
            const oneHot = inputs.one_hot[0];                   // Already 1D
            
            console.log('📊 JS Input data:');
            console.log(`  Audio features length: ${audioFeatures.length}`);
            console.log(`  Template length: ${template.length}`);
            console.log(`  One-hot length: ${oneHot.length}`);
            
            // Run the model for a single step
            const results = await this.runSingleInferenceStep(generator, audioFeatures, template, oneHot);
            
            this.jsResults = {
                outputs: results,
                shapes: {
                    new_vertice_out: [1, 1, results.new_vertice_out.length],
                    updated_vertice_emb: [1, 1, results.updated_vertice_emb.length]
                },
                model_type: "JavaScript FaceFormer Minimal"
            };
            
            console.log('✅ JavaScript model completed');
            console.log('📐 JS Output shapes:');
            Object.entries(this.jsResults.shapes).forEach(([key, shape]) => {
                console.log(`  ${key}: [${shape.join(', ')}]`);
            });
            
            return true;
            
        } catch (error) {
            console.error('❌ JavaScript model failed:', error);
            return false;
        }
    }

    async runSingleInferenceStep(generator, audioFeatures, template, oneHot) {
        // Run a single inference step using the minimal model
        console.log('🔄 Running single inference step...');
        
        try {
            // Create tensors for ONNX model
            const batchSize = 1;
            
            // Prepare inputs in correct shapes
            const audioTensor = new ort.Tensor('float32', new Float32Array(audioFeatures), [batchSize, 1, 768]);
            const templateTensor = new ort.Tensor('float32', new Float32Array(template), [batchSize, 1, 15069]);
            const oneHotTensor = new ort.Tensor('float32', new Float32Array(oneHot), [batchSize, 3]);
            
            // Initialize vertex embedding with small values
            const initialVerticeEmb = new Array(64).fill(0.01);
            const verticeEmbTensor = new ort.Tensor('float32', new Float32Array(initialVerticeEmb), [batchSize, 1, 64]);
            
            // Prepare model inputs
            const feeds = {
                audio_features: audioTensor,
                vertice_emb: verticeEmbTensor,
                one_hot: oneHotTensor,
                template: templateTensor
            };
            
            console.log('📋 Running ONNX inference...');
            const modelResults = await generator.session.run(feeds);
            
            // Extract results
            const newVerticeOut = Array.from(modelResults.new_vertice_out.data);
            const updatedVerticeEmb = Array.from(modelResults.updated_vertice_emb.data);
            
            console.log(`✅ Single step completed - Generated ${newVerticeOut.length} vertices`);
            
            return {
                new_vertice_out: newVerticeOut,
                updated_vertice_emb: updatedVerticeEmb,
                raw_outputs: {
                    new_vertice_out_tensor: modelResults.new_vertice_out,
                    updated_vertice_emb_tensor: modelResults.updated_vertice_emb
                }
            };
            
        } catch (error) {
            console.error('❌ Single inference step failed:', error);
            throw error;
        }
    }

    compareOutputs() {
        if (!this.pythonResults || !this.jsResults) {
            console.error('❌ Missing results for comparison');
            return null;
        }

        console.log('🔍 Comparing Python vs JavaScript outputs...');
        
        // Get outputs to compare
        const pythonOutputs = this.pythonResults.simplified_results?.simplified_python_outputs;
        const jsOutputs = this.jsResults.outputs;
        
        if (!pythonOutputs || !jsOutputs) {
            console.error('❌ Missing output data for comparison');
            return null;
        }

        // Compare shapes
        const shapeComparison = this.compareShapes();
        
        // Compare values
        const valueComparison = this.compareValues(pythonOutputs, jsOutputs);
        
        // Statistical analysis
        const stats = this.computeStatistics(pythonOutputs, jsOutputs);
        
        const comparison = {
            timestamp: new Date().toISOString(),
            shapes: shapeComparison,
            values: valueComparison,
            statistics: stats,
            summary: this.generateSummary(shapeComparison, valueComparison, stats)
        };
        
        this.displayComparison(comparison);
        return comparison;
    }

    compareShapes() {
        const pythonShapes = this.pythonResults.simplified_results?.shapes;
        const jsShapes = this.jsResults.shapes;
        
        const comparison = {};
        
        // Compare new_vertice_out shapes
        if (pythonShapes?.new_vertices && jsShapes?.new_vertice_out) {
            comparison.new_vertice_out = {
                python: pythonShapes.new_vertices,
                js: jsShapes.new_vertice_out,
                match: JSON.stringify(pythonShapes.new_vertices) === JSON.stringify(jsShapes.new_vertice_out)
            };
        }
        
        // Compare updated_vertice_emb shapes  
        if (pythonShapes?.updated_emb && jsShapes?.updated_vertice_emb) {
            comparison.updated_vertice_emb = {
                python: pythonShapes.updated_emb,
                js: jsShapes.updated_vertice_emb,
                match: JSON.stringify(pythonShapes.updated_emb) === JSON.stringify(jsShapes.updated_vertice_emb)
            };
        }
        
        return comparison;
    }

    compareValues(pythonOutputs, jsOutputs) {
        const comparison = {};
        
        // Compare new_vertice_out values
        if (pythonOutputs.new_vertice_out && jsOutputs.new_vertice_out) {
            comparison.new_vertice_out = this.compareArrays(
                pythonOutputs.new_vertice_out, 
                jsOutputs.new_vertice_out,
                'New Vertices'
            );
        }
        
        // Compare updated_vertice_emb values
        if (pythonOutputs.updated_vertice_emb && jsOutputs.updated_vertice_emb) {
            comparison.updated_vertice_emb = this.compareArrays(
                pythonOutputs.updated_vertice_emb,
                jsOutputs.updated_vertice_emb, 
                'Updated Embedding'
            );
        }
        
        return comparison;
    }

    compareArrays(array1, array2, name) {
        if (array1.length !== array2.length) {
            return {
                error: `Length mismatch: ${array1.length} vs ${array2.length}`,
                name: name
            };
        }
        
        const diffs = [];
        let sumSquaredDiff = 0;
        let maxDiff = 0;
        let totalValues = array1.length;
        
        for (let i = 0; i < array1.length; i++) {
            const diff = Math.abs(array1[i] - array2[i]);
            diffs.push(diff);
            sumSquaredDiff += diff * diff;
            maxDiff = Math.max(maxDiff, diff);
        }
        
        const meanSquaredError = sumSquaredDiff / totalValues;
        const rootMeanSquaredError = Math.sqrt(meanSquaredError);
        const meanAbsoluteError = diffs.reduce((a, b) => a + b, 0) / totalValues;
        
        return {
            name: name,
            length: totalValues,
            max_difference: maxDiff,
            mean_absolute_error: meanAbsoluteError,
            root_mean_squared_error: rootMeanSquaredError,
            first_10_diffs: diffs.slice(0, 10),
            is_close: maxDiff < 1e-3 && meanAbsoluteError < 1e-4
        };
    }

    computeStatistics(pythonOutputs, jsOutputs) {
        const stats = {};
        
        // Compute statistics for each output
        ['new_vertice_out', 'updated_vertice_emb'].forEach(key => {
            if (pythonOutputs[key] && jsOutputs[key]) {
                const pythonArray = pythonOutputs[key];
                const jsArray = jsOutputs[key];
                
                stats[key] = {
                    python: this.arrayStats(pythonArray),
                    js: this.arrayStats(jsArray)
                };
            }
        });
        
        return stats;
    }

    arrayStats(array) {
        const sum = array.reduce((a, b) => a + b, 0);
        const mean = sum / array.length;
        const variance = array.reduce((a, b) => a + (b - mean) ** 2, 0) / array.length;
        const std = Math.sqrt(variance);
        const min = Math.min(...array);
        const max = Math.max(...array);
        
        return {
            mean: mean,
            std: std,
            min: min,
            max: max,
            sum: sum
        };
    }

    generateSummary(shapeComparison, valueComparison, stats) {
        const summary = {
            shapes_match: true,
            values_close: true,
            issues: []
        };
        
        // Check shape matches
        Object.entries(shapeComparison).forEach(([key, comp]) => {
            if (!comp.match) {
                summary.shapes_match = false;
                summary.issues.push(`Shape mismatch in ${key}: Python ${JSON.stringify(comp.python)} vs JS ${JSON.stringify(comp.js)}`);
            }
        });
        
        // Check value closeness
        Object.entries(valueComparison).forEach(([key, comp]) => {
            if (comp.error) {
                summary.values_close = false;
                summary.issues.push(`${key}: ${comp.error}`);
            } else if (!comp.is_close) {
                summary.values_close = false;
                summary.issues.push(`${key}: Values not close enough (max_diff=${comp.max_difference.toFixed(6)}, mae=${comp.mean_absolute_error.toFixed(6)})`);
            }
        });
        
        // Overall assessment
        if (summary.shapes_match && summary.values_close) {
            summary.overall = "✅ MODELS MATCH - Outputs are consistent!";
        } else if (summary.shapes_match) {
            summary.overall = "⚠️ PARTIAL MATCH - Shapes match but values differ";
        } else {
            summary.overall = "❌ MISMATCH - Significant differences detected";
        }
        
        return summary;
    }

    displayComparison(comparison) {
        console.log('\n' + '='.repeat(60));
        console.log('🔍 FACEFORMER MODEL COMPARISON RESULTS');
        console.log('='.repeat(60));
        
        console.log('\n📐 Shape Comparison:');
        Object.entries(comparison.shapes).forEach(([key, comp]) => {
            const status = comp.match ? '✅' : '❌';
            console.log(`  ${status} ${key}:`);
            console.log(`    Python: [${comp.python.join(', ')}]`);
            console.log(`    JS:     [${comp.js.join(', ')}]`);
        });
        
        console.log('\n📊 Value Comparison:');
        Object.entries(comparison.values).forEach(([key, comp]) => {
            if (comp.error) {
                console.log(`  ❌ ${key}: ${comp.error}`);
            } else {
                const status = comp.is_close ? '✅' : '⚠️';
                console.log(`  ${status} ${comp.name}:`);
                console.log(`    Max Difference: ${comp.max_difference.toFixed(6)}`);
                console.log(`    Mean Absolute Error: ${comp.mean_absolute_error.toFixed(6)}`);
                console.log(`    Root Mean Squared Error: ${comp.root_mean_squared_error.toFixed(6)}`);
            }
        });
        
        console.log('\n📈 Statistics Summary:');
        Object.entries(comparison.statistics).forEach(([key, stats]) => {
            console.log(`  ${key}:`);
            console.log(`    Python - Mean: ${stats.python.mean.toFixed(4)}, Std: ${stats.python.std.toFixed(4)}, Range: [${stats.python.min.toFixed(4)}, ${stats.python.max.toFixed(4)}]`);
            console.log(`    JS     - Mean: ${stats.js.mean.toFixed(4)}, Std: ${stats.js.std.toFixed(4)}, Range: [${stats.js.min.toFixed(4)}, ${stats.js.max.toFixed(4)}]`);
        });
        
        console.log('\n🎯 Summary:');
        console.log(`  ${comparison.summary.overall}`);
        if (comparison.summary.issues.length > 0) {
            console.log('\n⚠️ Issues Found:');
            comparison.summary.issues.forEach((issue, i) => {
                console.log(`  ${i + 1}. ${issue}`);
            });
        }
        
        console.log('\n' + '='.repeat(60));
    }

    async saveComparisonResults(comparison) {
        const data = {
            comparison: comparison,
            python_results: this.pythonResults,
            js_results: this.jsResults,
            timestamp: new Date().toISOString()
        };
        
        // For browser environment, we'll log the data
        // In a real environment, you'd save this to a file
        console.log('💾 Full comparison data:', data);
        
        // Also create a downloadable JSON
        const blob = new Blob([JSON.stringify(data, null, 2)], {type: 'application/json'});
        const url = URL.createObjectURL(blob);
        
        const a = document.createElement('a');
        a.href = url;
        a.download = 'faceformer_comparison_results.json';
        document.body.appendChild(a);
        a.click();
        document.body.removeChild(a);
        URL.revokeObjectURL(url);
        
        console.log('💾 Comparison results downloaded as JSON file');
    }

    async runFullComparison() {
        console.log('🚀 Starting Full FaceFormer Model Comparison...');
        
        // Step 1: Load Python results
        const pythonLoaded = await this.loadPythonResults();
        if (!pythonLoaded) {
            console.error('❌ Cannot proceed without Python results');
            return false;
        }
        
        // Step 2: Run JavaScript model
        const jsRan = await this.runJavaScriptModel();
        if (!jsRan) {
            console.error('❌ Cannot proceed without JavaScript results');
            return false;
        }
        
        // Step 3: Compare outputs
        const comparison = this.compareOutputs();
        if (!comparison) {
            console.error('❌ Comparison failed');
            return false;
        }
        
        // Step 4: Save results
        await this.saveComparisonResults(comparison);
        
        console.log('✅ Full comparison completed successfully!');
        return true;
    }
}

// Export for use
if (typeof module !== 'undefined' && module.exports) {
    module.exports = { FaceFormerComparison };
}

// Make available globally for web use
if (typeof window !== 'undefined') {
    window.FaceFormerComparison = FaceFormerComparison;
}
